ddtrace/sampler.py

```diff
@@ -92,3 +103,251 @@ def set_sample_rate_by_service(self, rate_by_service):
 RateByServiceSampler._default_key = RateByServiceSampler._key()
+
+
+class DatadogSampler(BaseSampler):
+    """
+    """
+    __slots__ = ('rules', 'rate_limit')
+
+    DEFAULT_RATE_LIMIT = 100
+    NO_RATE_LIMIT = -1
+
+    def __init__(self, rules=None, rate_limit=None):
+        """
+        Constructor for DatadogSampler sampler
+
+        :param rules: List of :class:`SamplingRule` rules to apply to the root span of every trace, default no rules
+        :type rules: :obj:`list` of :class:`SamplingRule`
+        :param rate_limit: Global rate limit (traces per second) to apply to all traces regardless of the rules
+            applied to them, default 100 traces per second
+        :type rate_limit: :obj:`int`
+        """
+        # Ensure rules is a list
+        if not rules:
+            rules = []
+
+        # Validate that the rules is a list of SampleRules
+        for rule in rules:
+            if not isinstance(rule, SamplingRule):
+                raise TypeError('Rule {!r} must be a sub-class of type ddtrace.sampler.SamplingRules'.format(rule))
+        self.rules = rules
+
+        # Configure rate limiter
+        if rate_limit is None:
+            rate_limit = self.DEFAULT_RATE_LIMIT
+        self.limiter = RateLimiter(rate_limit)
+
+    def sample(self, span):
+        """
+        Decide whether the provided span should be sampled or not
+
+        The span provided should be the root span in the trace.
+
+        :param span: The root span of a trace
+        :type span: :class:`ddtrace.span.Span`
+        :returns: Whether the span was sampled or not
+        :rtype: :obj:`bool`
+        """
+        # If there are rules defined, then iterate through them and find one that wants to sample
+        matching_rule = None
+        if self.rules:
+            # Go through all rules and grab the first one that matched
+            # DEV: This means rules should be ordered by the user from most specific to least specific
+            for rule in self.rules:
+                if rule.matches(span):
+                    matching_rule = rule
+                    break
+            else:
+                # No rule matched, reject
+                if span._context:
+                    span._context.sampling_priority = AUTO_REJECT
+                return False
+
+            if not matching_rule.sample(span):
+                if span._context:
+                    span._context.sampling_priority = AUTO_REJECT
+                return False
+            else:
+                # Do not return here, we need to apply rate limit
+                if span._context:
+                    span._context.sampling_priority = AUTO_KEEP
+
+        # Ensure all allowed traces adhere to the global rate limit
+        if not self.limiter.is_allowed():
+            if span._context:
+                span._context.sampling_priority = AUTO_REJECT
+            return False
+
+        # TODO: Set `_sampling_priority_rate_v1` tag based on
```

Comment:
@LotharSee You mentioned in the RFC that we should also send metrics for each type of keep/drop reason. Does that mean we should keep applied rates for each metric as well?